posts/unified-memory-oversubscription/uvm_oversubs.cu

```diff
@@ -0,0 +1,501 @@
+      {
+        // distribute pages across GPU0 and CPU
+        // get page-split ratios
+        float cpu_factor = oversubscription_factor - 1.0;
+        if (cpu_factor < 0.0f)
+          cpu_factor = 0.0f; 
+        int mod_zero_devId = cudaCpuDeviceId;
+        int flip_devId = current_device;
+        int mod_scale = num_pages;
+        if (cpu_factor > 1.0) {
+          mod_zero_devId = current_device;
+          flip_devId = cudaCpuDeviceId;
+          mod_scale = int(std::round(oversubscription_factor));
+        }
+        else if (cpu_factor != 0.0f) {
+          mod_scale = int(std::round(oversubscription_factor / cpu_factor));
+        }
+        int gpu_page_count = 0, cpu_page_count = 0;
+        void *running_ptr = uvm_alloc_ptr;
+      
+        for (int i = 0; i < num_pages; i++) {
+          int device = flip_devId;
+          if ((i % mod_scale) == 0 && i != 0)
+            device = mod_zero_devId;
+
+          if (gpu_page_count == permissible_phys_pages_count)
+            device = cudaCpuDeviceId;
+
+          CUDA_CHECK(cudaMemPrefetchAsync(running_ptr, page_size, device, task_stream));
+
+          if (device == cudaCpuDeviceId)
+            cpu_page_count++;
+          else
+            gpu_page_count++;
+
+          if (itr == 0) {
+            CUDA_CHECK(cudaMemAdvise(running_ptr, page_size, cudaMemAdviseSetPreferredLocation,
+                                      device));
+
+            if (device == cudaCpuDeviceId)
+              CUDA_CHECK(cudaMemAdvise(running_ptr, page_size, cudaMemAdviseSetAccessedBy,
+                                        current_device));
+          }
+          running_ptr = reinterpret_cast<void*>((size_t)running_ptr + page_size);
+        }
+      }
+      break;
+      case PREFETCH_ONCE_AND_HINTS:
+      {
+      // in oversubscrription this is going to over-flow back to sysmem
+        if (itr == 0) {
+          CUDA_CHECK(cudaMemAdvise(uvm_alloc_ptr, allocation_size, cudaMemAdviseSetAccessedBy,
+                                    current_device));
+          CUDA_CHECK(cudaMemPrefetchAsync(uvm_alloc_ptr, allocation_size, current_device,
+                                            task_stream));
+        }
+      }
+      break;
+      case ZERO_COPY:
+      {
+        if (itr == 0) {
+          CUDA_CHECK(cudaMemAdvise(uvm_alloc_ptr, allocation_size,
+                                    cudaMemAdviseSetPreferredLocation, cudaCpuDeviceId));
+          CUDA_CHECK(cudaMemAdvise(uvm_alloc_ptr, allocation_size,
+                                    cudaMemAdviseSetAccessedBy, current_device));
+        }
+      }
+      default:
+        break;
+      }
+
+      //CUDA_CHECK(cudaDeviceSynchronize());
+      // timer start
+      CUDA_CHECK(cudaEventRecord(startE, task_stream));
+
+      // run read/write kernel for streaming/random access
+      if (k_op == READ) {
+        if (memory_access == STREAMING) {
+          cudaOccupancyMaxActiveBlocksPerMultiprocessor(&num_blocks_per_sm,
```

Comment:
Should we move `cudaOccupancyMaxActiveBlocksPerMultiprocessor` and assigning `grid.x` from the critical path here? It can be done once in the beginning to determine the launch config. I suppose it doesn't change perf, but will make code a little cleaner.

Reply:
done